File: native/ios/VocaKit/Sources/VocaSession/Cloud/PeerLeg.swift
import Foundation
import VocaKit

// MARK: - Peer connection abstraction
//
// One "leg" is one RTCPeerConnection plus the microphone clone that feeds it
// (CONTRACT §3 step 2–5). The view model only ever needs five things from a
// leg, so that is the whole protocol; `WebRTCLeg` is the real thing and the
// tests use an in-memory fake, which is how the state machine gets exercised
// without ever touching libwebrtc.

public enum PeerConnectionState: Equatable, Sendable {
    case new
    case connecting
    case connected
    case disconnected
    case failed
    case closed
}

public struct LocalOffer: Equatable, Sendable {
    public let sdp: String
    public let type: String

    public init(sdp: String, type: String = "offer") {
        self.sdp = sdp
        self.type = type
    }
}

@MainActor
public protocol PeerLeg: AnyObject {
    /// Fired on the main actor whenever the underlying connection state changes.
    var onConnectionState: ((PeerConnectionState) -> Void)? { get set }

    /// Add the (enabled) mic clone, create + set the local offer, wait for ICE
    /// gathering to finish (capped at 5s) and return what to POST to `/offer`.
    func createOffer() async throws -> LocalOffer

    /// Apply the server's answer.
    func setRemoteAnswer(sdp: String, type: String) async throws

    /// Mute/unmute this leg's mic clone. Push-to-talk flips this.
    func setTrackEnabled(_ enabled: Bool)

    /// Stop the local track, drop the remote audio, close the connection.
    func close()
}

@MainActor
public protocol PeerLegFactory {
    func makeLeg(iceServers: [ICEServer]) throws -> PeerLeg
}

// MARK: - Microphone permission

/// The one bit of AVFoundation the view model needs before anything else
/// (CONTRACT §3 step 1). Injected so tests can say "denied".
public protocol MicrophonePermission: Sendable {
    func request() async -> Bool
}

public enum CloudSessionError: Error, Equatable, Sendable {
    case notConnected
    case noIceServers
    case connectionLost
}

File: native/android/session/src/main/kotlin/com/vocare/translate/session/cloud/PeerLeg.kt
package com.vocare.translate.session.cloud

import com.vocare.translate.core.model.IceServer

/** Mirror of the subset of `RTCPeerConnectionState` the session cares about. */
enum class LegConnectionState { NEW, CONNECTING, CONNECTED, DISCONNECTED, FAILED, CLOSED }

/** A local session description as the backend wants it. */
data class LocalOffer(val sdp: String, val type: String = "offer")

/**
 * One participant leg: a peer connection carrying a clone of the microphone up
 * and the translation for that person back down.
 *
 * `WebRtcLeg` is the real thing; tests substitute a fake so the view model's
 * protocol (CONTRACT §3) can be exercised without libwebrtc.
 */
interface PeerLeg {
    /** Turn this leg's microphone clone on or off (push-to-talk owns this once connected). */
    fun setMicEnabled(enabled: Boolean)

    /**
     * Add the mic track, create the offer, set it locally and wait for ICE gathering
     * to complete (capped at 5s, JSX `waitForICE`). Returns the local description.
     */
    suspend fun createOffer(): LocalOffer

    suspend fun setAnswer(sdp: String, type: String)

    /** Stop the track and close the connection. Idempotent. */
    fun close()
}

/** Creates legs; both legs of a session share the same ICE servers and microphone. */
fun interface PeerLegFactory {
    /**
     * @param onConnectionState called (on any thread) whenever the leg's connection state changes.
     */
    fun create(iceServers: List<IceServer>, onConnectionState: (LegConnectionState) -> Unit): PeerLeg
}
